packages/block-editor/src/store/actions.js

```diff
@@ -1238,6 +1238,24 @@ export function removeBlock( clientId, selectPrevious ) {
 	return removeBlocks( [ clientId ], selectPrevious );
 }
 
+/**
+ * Returns an action object used in signalling that a removal prompt must be displayed.
+ *
+ * @param {boolean}  displayPrompt Whether to prompt for removal.
+ * @param {Function} options       Function to call if removal is confirmed and blockName.
+ *
+ * @return {Object} Action object.
+ */
+export function displayRemovalPrompt( displayPrompt, options = {} ) {
+	const { removalFunction, blockName } = options;
+	return {
+		type: 'PROMPT_REMOVAL',
+		displayPrompt,
+		removalFunction,
+		blockName,
+	};
+}
```

Comment:
I wonder if we can / want to fold this logic into the [`removeBlocks` action](https://github.com/WordPress/gutenberg/blob/549ab383905d35b3062f4c76d856ce6937cb646a/packages/block-editor/src/store/actions.js#L1198C14-L1225), since that's the central code path for removal **and** it already is the site for checks and conditional dispatching:

- checks for `canRemoveBlocks` before proceeding
- dispatches `selectPreviousBlock` alongside the removal action if requested
- ensures correct focus by dispatching `ensureDefaultBlock` at the very end

I'm not sure of this idea, so I'd like to be challenged, but what about something like:

```diff
export const removeBlocks =
-	( clientIds, selectPrevious = true ) =>
+	( clientIds, selectPrevious = true, forceRemove = false ) =>
	( { select, dispatch } ) => {
		...
		if ( ! canRemoveBlocks ) {
			return;
		}

+		if ( ! forceRemove && needsConsentForRemoval( blockIds ) ) {
+			dispatch( { type: 'PROMPT_FOR_BLOCK_REMOVAL', blockIds } );
+			return;
+		}

		...
		dispatch( { type: 'REMOVE_BLOCKS', clientIds } );
+		dispatch( { type: 'PROMPT_FOR_BLOCK_REMOVAL', null } );
		...
	};
```

then the modal would dispatch a new remove action with `forceRemove: true`

Reply:
Yeah, I thought about that (see my little monologue in the PR description 😄) but hesitate to do it because `removeBlocks` gets called in a lot of places (itself or via `removeBlock`) where the prompt isn't relevant, and that would mean going through the logic to check if there's a critical block in the list of blocks to remove every time. 

It does feel conceptually tidier to have everything together in removeBlocks though.

I'm also happy to hear more thoughts about this!